optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/tri/TriConstraintStream.java

```diff
@@ -688,6 +688,106 @@
+                    TriFunction<A, B, C, GroupKeyC_> groupKeyCMapping,
+                    TriConstraintCollector<A, B, C, ResultContainerD_, ResultD_> collectorD);
+
+    /**
+     * Convert the {@link TriConstraintStream} to a {@link QuadConstraintStream}, consisting of unique tuples with four
+     * facts.
+     * <p>
+     * The first fact is the return value of the first group key mapping function, applied on the incoming tuple.
+     * The second fact is the return value of the second group key mapping function, applied on all incoming tuples with
+     * the same first fact.
+     * The third fact is the return value of the third group key mapping function, applied on all incoming tuples with
+     * the same first fact.
+     * The fourth fact is the return value of the fourth group key mapping function, applied on all incoming tuples with
+     * the same first fact.
+     *
+     * @param groupKeyAMapping never null, function to convert the original tuple into a first fact
+     * @param groupKeyBMapping never null, function to convert the original tuple into a second fact
+     * @param groupKeyCMapping never null, function to convert the original tuple into a third fact
+     * @param groupKeyDMapping never null, function to convert the original tuple into a fourth fact
+     * @param <GroupKeyA_> the type of the first fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyB_> the type of the second fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyC_> the type of the third fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyD_> the type of the fourth fact in the destination {@link QuadConstraintStream}'s tuple
+     * @return never null
+     */
+    <GroupKeyA_, GroupKeyB_, GroupKeyC_, GroupKeyD_>
+            QuadConstraintStream<GroupKeyA_, GroupKeyB_, GroupKeyC_, GroupKeyD_> groupBy(
+                    TriFunction<A, B, C, GroupKeyA_> groupKeyAMapping, TriFunction<A, B, C, GroupKeyB_> groupKeyBMapping,
+                    TriFunction<A, B, C, GroupKeyC_> groupKeyCMapping, TriFunction<A, B, C, GroupKeyD_> groupKeyDMapping);
+
+    // ************************************************************************
+    // Operations w/ duplicate tuple possibility
+    // ************************************************************************
+
+    /**
+     * As defined by {@link UniConstraintStream#map(Function)}.
+     *
+     * @param mapping never null, function to convert the original tuple into the new tuple
+     * @param <ResultA_> the type of the only fact in the resulting {@link UniConstraintStream}'s tuple
+     * @return never null
+     */
+    <ResultA_> UniConstraintStream<ResultA_> map(TriFunction<A, B, C, ResultA_> mapping);
+
+    /**
+     * Transforms the stream in such a way that all the tuples going through it are distinct.
```

Comment:
"in such a way" seems a bit verbose, how about something like:

```suggestion
     * Removes duplicate tuples from the stream, according to the tuple's fact's {@link Object#equals(Object)equals}/hashcode methods, such that only distinct tuples remain.
```